wandb/sdk/lib/redirect.py

```diff
@@ -1,301 +1,624 @@
-
-    def _stop(self):
-        name = self._name
-        logger.info("_stop: %s", name)
+        # self._prev_callback_timestamp = time.time()
+        self._pipe_relay_thread = threading.Thread(target=self._pipe_relay)
+        self._pipe_relay_thread.daemon = True
+        self._pipe_relay_thread.start()
+        if not wandb.run or wandb.run._settings.mode == "online":
+            self._callback_thread = threading.Thread(target=self._callback)
+            self._callback_thread.daemon = True
+            self._callback_thread.start()
 
+    def uninstall(self):
+        if not self._installed:
+            return
+        self._installed = False
+        self.src_wrapped_stream.flush()
+        time.sleep(1)
         self._stopped.set()
-        os.write(self._pipe_wr, _LAST_WRITE_TOKEN.encode())
-        os.close(self._pipe_wr)
-
-        logger.info("_stop closed: %s", name)
-        # TODO: need to shut this down cleanly since it is a daemon thread
-        self._thread.join(timeout=30)
-        if self._thread.is_alive():
-            logger.error("Thread did not join: %s", self._name)
-            # TODO(jhr): do something better
-        logger.info("_stop joined: %s", name)
-        os.close(self._pipe_rd)
-        logger.info("_stop rd closed: %s", name)
+        os.dup2(self._orig_src_fd, self.src_fd)
+        os.write(self._pipe_write_fd, b"\n")
+        os.close(self._pipe_write_fd)
+        os.close(self._pipe_read_fd)
+        self.flush()
+        _WSCH.remove_fd(self._pipe_read_fd)
+        super(Redirect, self).uninstall()
+
+    def flush(self):
+        data = self._emulator.read().encode("utf-8")
+        if data:
+            for cb in self.cbs:
+                try:
+                    cb(data)
+                except Exception:
+                    pass  # TODO(frz)
+
+    def _callback(self):
+        while not self._stopped.is_set():
+            # if time.time() - self._prev_callback_timestamp < _MIN_CALLBACK_INTERVAL:
+            #     time.sleep(0.1)
+            #     continue
+            self.flush()
+            time.sleep(_MIN_CALLBACK_INTERVAL)
+            # self._prev_callback_timestamp = time.time()
+
+    def _pipe_relay(self):
+        while True:
+            try:
+                data = os.read(self._pipe_read_fd, 4096)
+                if self._stopped.is_set():
+                    return
+            except OSError:
+                return
+            try:
+                i = self._orig_src.write(data)
+                if i is not None:  # python 3 w/ unbuffered i/o: we need to keep writing
+                    while i < len(data):
+                        i += self._orig_src.write(data[i:])
+                self.src_wrapped_stream.flush()
+            except OSError:
+                return
+            try:
+                self._emulator.write(data.decode("utf-8"))
+            except UnicodeDecodeError:
+                pass  # TODO(frz): partial unicode character?
```

Comment:
This is a BEAST.  It's looking good, let's add some doc strings to describe what these various classes and methods are doing.